verifier/src/evidence.rs

```diff
@@ -199,47 +203,133 @@ where
             certificate_verifier,
             _trusted_identities: trusted_identities.into_iter().map(Into::into).collect(),
             time,
+            tcb_verifier: Cell::new(None),
         }
     }
 
-    fn verify_tcb_signing_key(
+    fn verify_certificate_chain<'c, 'd>(
+        &self,
+        chain: &[Certificate],
+        crls: impl IntoIterator<Item = &'d CertificateList>,
+    ) -> (
+        VerifyingKey,
+        VerificationOutput<Option<CertificateChainVerifierError>>,
+    ) {
+        let result = self
+            .certificate_verifier
+            .verify_certificate_chain(chain, crls, self.time);
+        let is_success = result.is_ok() as u8;
+
+        // Using the default key will result in the user seeing "Error verifying the signature" for
+        // the signed data. So we try to get the key from the certificate chain, even if the
+        // verification failed. This handles the most likely failure case of an expired
+        // certificate, whose key is still the key that signed the data of interest.
+        let key = match chain.first() {
```

Comment:
oh yeah that is much nicer! Especially once `key_from_certificate` has been updated to an `Option`. 

Updated.

> Unclear if that is actually more readable...

I've been trying to get more familiar with the functional chaining style, it can get a bit dense at times but I think for this case it should be easier to grok once one becomes more accustomed to the `a -> b -> c` of `foo.a().b().c()`